### DISCUSSION TITLE
Does Renovate support excluding digest updates on NPM?

### DISCUSSION THREAD

**Discussion by arrrgi** [renovatebot/renovate#9139] [Request Help]
I have a recurring PR on hosted Gitlab for a monorepo dependency that has a digest version which was released without a tag. Renovate wants to keep creating PR's which I close but I'd like to avoid these being created for anything but major/minor/patch for the monorepo.

My config looks like this:

```json
{
  "extends": [
    ":automergeDisabled",
    ":dependencyDashboard",
    ":docker",
    ":ignoreModulesAndTests",
    ":ignoreUnstable",
    ":renovatePrefix",
    ":semanticCommits",
    ":semanticPrefixChore",
    ":separateMajorReleases",
    "docker:disableMajor",
    "group:monorepos",
    "group:recommended"
  ],
  "commitBodyTable": true,
  "dependencyDashboardAutoclose": true,
  "docker": {
    "pinDigests": true
  },
  "packageRules": [
    {
      "extends": "monorepo:docusaurus",
      "groupName": "docusaurus monorepo",
      "matchUpdateTypes": ["major", "minor", "patch"]
    }
  ],
  "enabledManagers": ["dockerfile", "npm"],
  "labels": ["renovate"],
  "lockFileMaintenance": {
    "enabled": true
  },
  "node": {
    "supportPolicy": ["lts_latest"]
  },
  "prCreation": "status-success",
  "rangeStrategy": "bump",
  "separateMinorPatch": true,
  "schedule": "every weekend",
  "stabilityDays": 3
}
```

My understanding is that versions like this **2.0.0-alpha.71** should match, while versions like this **v2.0.0-alpha.ffe8b6106** should be ignored.

Does Renovate classify a digest tagged update as a patch? Am I missing something from my config? Is the pinDigests rule being interpreted globally instead of just for Dockerfiles?

**Comment by rarkins** [comment]
`v2.0.0-alpha.ffe8b6106` is a version, not a digest. It just so happens that the publishers are choosing to put a digest into the version string. Assuming this is related to Docusaurus, @HonkingGoose I think you already have a solution to this for your own needs? I forget where we discussed it, but assume you have it handy

**Reply by HonkingGoose** [reply]
NOTE: deleted my previous reply to copy/paste it below, as you cannot mark a "reply" as an answer in a GitHub discussion...

**Comment by HonkingGoose** [comment]
> Assuming this is related to Docusaurus, @HonkingGoose I think you already have a solution to this for your own needs? I forget where we discussed it, but assume you have it handy

Good memory! 😄 Docusaurus uses some weird versioning unfortunately.
Luckily I already got config help from the Renovate wizards. 🧙 🧙 🙃 

I'm using this config to filter out the `canary` releases and let through the normal full-fat releases:

```json
  "packageRules": [
    {
      "description": "Allow mainline @docusaurus releases, ignore canary releases",
      "matchPackagePatterns": ["^@docusaurus"],
      "allowedVersions": "/^\\d+\\.\\d+\\.\\d+(-alpha\\.(\\d+))?$/"
    }
```

If you want to see the config in the context of a complete Renovate config, you can check out: https://github.com/HonkingGoose/git-gosling/blob/main/renovate.json to see the full config I'm using now.

**Reply by arrrgi** [reply]
Thanks @rarkins for clearing that up and @HonkingGoose for the answer. I suspected I might have to get down and dirty with some regex but wanted to make sure I hadn't missed something

**Reply by arrrgi** [reply]
I tried to get a little "clever" with these rules but found I was still getting the `canary` releases 

```
"packageRules": [
    {
      "description": "Allow mainline @docusaurus releases, ignore canary releases",
      "groupName": "docusaurus monorepo",
      "matchSourceUrlPrefixes": ["https://github.com/facebook/docusaurus"],
      "matchUpdateTypes": ["major", "minor", "patch"],
      "allowedVersions": "!/\\b([a-f0-9]{9})?$/"
    }
  ]
```

Would `matchUpdateTypes` be overriding the `allowedVersions` rule which I've configured as an exclusion filter?

**Reply by rarkins** [reply]
You can't mix `matchUpdateTypes` with `allowedVersions` in the same package rule. It's something we should try to lint and warn about, but we don't yet. The reason is because versions filtering is done the step *before* we know the update type. So by the time you. know the update type and match the rule, it's too late to apply the `allowedVersions` rule. I don't think that extra line achieves anything though and can be removed.

**Reply by arrrgi** [reply]
Thanks, will remove that and confirm whether that changes the results

**Reply by arrrgi** [reply]
Worked like a charm! Thanks @rarkins 👍

**Reply by HonkingGoose** [reply]
@madcatsu Could you maybe copy/paste your new config? I'm curious what you ended up with, and if that is easier to understand/reason about than my current config?

**Reply by arrrgi** [reply]
Sure, my full config looks like this now:

```json
{
  "extends": [
    ":automergeDisabled",
    ":dependencyDashboard",
    ":docker",
    ":ignoreModulesAndTests",
    ":ignoreUnstable",
    ":renovatePrefix",
    ":semanticCommits",
    ":semanticPrefixChore",
    ":separateMajorReleases",
    "docker:disableMajor",
    "group:monorepos",
    "group:recommended"
  ],
  "assignees": ["@madcatsu"],
  "commitBodyTable": true,
  "dependencyDashboardAutoclose": true,
  "docker": {
    "pinDigests": true
  },
  "packageRules": [
    {
      "description": "Allow only @docusaurus mainstream releases",
      "groupName": "docusaurus monorepo",
      "matchSourceUrlPrefixes": ["https://github.com/facebook/docusaurus"],
      "allowedVersions": "!/\\b([a-f0-9]{9})?$/"
    }
  ],
  "enabledManagers": ["dockerfile", "npm"],
  "labels": ["renovate"],
  "lockFileMaintenance": {
    "enabled": true
  },
  "node": {
    "supportPolicy": ["lts_latest"]
  },
  "prCreation": "status-success",
  "prHourlyLimit": 0,
  "prConcurrentLimit": 0,
  "rangeStrategy": "bump",
  "separateMinorPatch": true,
  "schedule": "every weekend",
  "stabilityDays": 3,
  "timezone": "Australia/Brisbane",
  "vulnerabilityAlerts": {
    "labels": ["security"],
    "assignees": ["@madcatsu"]
  }
}
```

**Reply by HonkingGoose** [reply]
Cheers! 😄

**Reply by arrrgi** [reply]
@HonkingGoose - is that working for you? Renovate isn't picking up the 2.0.0-alpha.72 release for me even though that rule is an exclusion, not a selective inclusion rule.

**Reply by HonkingGoose** [reply]
@madcatsu I have not tested your new config yet, but you should have gotten a PR by now, as alpha.72 is more than 3 `stabilityDays` old. So something about your configuration is still not right, but I'm not sure what the problem is.

I got a update PR from Renovate bot for `alpha.72` just fine using the config below (See PR https://github.com/HonkingGoose/git-gosling/pull/166):

```json
  "packageRules": [
    {
      "description": "Allow mainline @docusaurus releases, ignore canary releases",
      "matchPackagePatterns": ["^@docusaurus"],
      "allowedVersions": "/^\\d+\\.\\d+\\.\\d+(-alpha\\.(\\d+))?$/"
    },
    {
      "description": "Wait 3 days before creating a npm update PR",
      "matchDatasources": ["npm"],
      "stabilityDays": 3
    }
  ],
```
